django/db/models/expressions.py

```diff
@@ -483,7 +485,14 @@ def __init__(self, value, output_field=None):
         self.value = value
 
     def as_sql(self, compiler, connection):
-        return '%s', [self.value]
+        val = self.value
+        if self._output_field_or_none is not None:
+            from django.db.models.sql.compiler import SQLUpdateCompiler
+            if isinstance(compiler, SQLUpdateCompiler):
```

Comment:
I've researched the ```Value.as_sql``` topic a bit.
The documentation definitely states that there is a difference in usage between ```get_db_prep_value``` and ```get_db_prep_save```.
https://docs.djangoproject.com/en/1.7/ref/models/fields/#django.db.models.Field.get_db_prep_save
https://docs.djangoproject.com/en/1.7/howto/custom-model-fields/#converting-query-values-to-database-values

Browsing the code, I think ```DecimalField``` highlights the difference the best. For saving, the value is converted to a ```Decimal``` with the required precision. For querying, it remains unchanged, which makes sense since when querying a ```DecimalField(decimal_places=0)``` a value of ```0.5``` should be accepted as is, but should not match anything.

So in my opinion this needs to stay, but needs to be delegated to the compiler class.
I'm also thinking that for SELECT this should be ```get_db_prep_loopup('exact', value, connection)``` instead of ```get_db_prep_value(value, connection)```.